gh-58451: Add optional delete_on_close parameter to NamedTemporaryFile

Ev2geny opened the pull request:
This is a reincarnation of the already marked as **awaiting merge** [PR 22431](https://github.com/python/cpython/pull/22431), which was subsequently closed due to some [ git mistake](https://github.com/python/cpython/pull/22431#issuecomment-1250404118).

This PR fixes issue [58451](https://github.com/python/cpython/issues/58451): **tempfile.NamedTemporaryFile not particularly useful on Windows**

```tempfile.NamedTemporaryFile``` is too hard to use portably when you need to open the file by name after writing it.  To do that, you need to close the file first (on Windows), which means you have to pass ```delete=False```, which in turn means that you get no help in cleaning up the actual file resource,

Hence at the moment there is no out of the box solution to use  tempfile.NamedTemporaryFile on Windows in such scenario (which is often used in unit testing):

* in test module:
1) create and open temporary file
2) write data to it
3) pass name of the temporary file to the operational code

* In operational code, being tested
1) open file, using name of the temporary file 
2) read data from this temporary file 

In this Pull Request the issue is solved by adding an additional optional argument to ```NamedTemporaryFile() 'delete_on_close' ``` (default is True). It works in combination with already existing argument ```'delete'```, and determines the deletion behaviour.

If _delete_ is true (the default) and _delete_on_close_ is true (the default), the file is deleted as soon as it is closed. If _delete_ is true and _delete_on_close_ is false, the file is deleted on context manager exit only, if no context manager was used, then the file is deleted only when the file-like object is finalized.  If __delete__ is false, the value of _delete_on_close_ is ignored.

So, the change shall be fully backwards compatible.

<!-- gh-issue-number: gh-58451 -->
* Issue: gh-58451
<!-- /gh-issue-number -->

Ev2geny committed:
Brought the same changes as in now closed PR 22431, but to the latest version of main

Ev2geny committed:
Based on the review from @eryksun the code and the documentation is updated to handle the situation, when delete = True, delete_on_close = Flase and no context manager was used

Ev2geny committed:
documentation is slightly corrected

Ev2geny committed:
minor documentation update

merwok commented on Doc/whatsnew/3.12.rst:
```diff
@@ -142,6 +142,11 @@ unicodedata
+tempfile
+--------
+
+The :class:`tempfile.NamedTemporaryFile` class has a new optional parameter
```
```suggestion
The :class:`tempfile.NamedTemporaryFile` function has a new optional parameter
```

merwok commented on Misc/NEWS.d/next/Library/2020-09-28-04-56-04.bpo-14243.YECnxv.rst:
```diff
@@ -0,0 +1 @@
+The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close*
```
```suggestion
The :class:`tempfile.NamedTemporaryFile` function has a new optional parameter *delete_on_close*
```

We have to say either `Name does x and y` or `the Name module/class/function/attribute does x and y`

Ev2geny committed:
Update Doc/whatsnew/3.12.rst

Co-authored-by: Éric <merwok@netwok.org>

eryksun commented on Lib/tempfile.py:
```diff
@@ -518,18 +530,32 @@ def __iter__(self):
         for line in self.file:
             yield line
 
+    def __del__(self):
```
I'd add the class attributes `name = None` and `_cleanup_called = False`, and a common `_cleanup()` method. For example:

```python
    name = None
    _cleanup_called = False

    def _cleanup(self):
        if self._cleanup_called or self.name is None:
            return
        self._cleanup_called = True
        try:
            self.close()
        finally:
            # If the file is to be deleted, but not on close, delete it now.
            if self.delete and not self.delete_on_close:
                try:
                    _os.unlink(self.name)
                except FileNotFoundError:
                    # The file may have been deleted already.
                    pass

    def __exit__(self, exc, value, tb):
        result = self.file.__exit__(exc, value, tb)
        self._cleanup()
        return result

    def __del__(self):
        self._cleanup()
```

Ev2geny replied:
@eryksun , I am sorry if this is a stupid question, but can you please explain me why you use **class attributes** `name = None` and `_cleanup_called = False` ? Why can't they be an instance attributes:

```py
def __init__(self, .....)
    name = None
    _cleanup_called = False
```

Class attributes are normally used to represent data shared across all instances of the class. Surely we want to have these flags to be specific for every temporary file instance (or am I badly missing something?)

And then in this line
```py
    self._cleanup_called = True
```
you create already an instance attribute with the same name. So, what is the point?

I noticed, that the same trick is used in the `_TemporaryFileCloser`

```py
class _TemporaryFileCloser:
    """A separate object allowing proper closing of a temporary file's
    underlying file object, without adding a __del__ method to the
    temporary file."""

    file = None  # Set here since __del__ checks it
    close_called = False
```

eryksun replied:
> can you please explain me why you use class attributes

I copied the pattern from `_TemporaryFileCloser`. I guess in some cases it could be safer than relying only on the instance attribute. For example, if `self._closer` were assigned first in `__init__()`, and instantiating the closer failed, then the wrapper's `__del__()` finalizer would see that `self.name is None` (the default value set as a class attribute) and thus skip trying to delete the file. That leaves it up to `NamedTemporaryFile()` to delete the file, which to me seems the proper thing to do since instantiating the wrapper failed.

Also, in the case of attributes such as `close_called`, which usually have the default value for the normal lifetime of an instance, using a class attribute can reduce the memory footprint. This could be a significant benefit if there are thousands of instances, but generally it's not a concern.

eryksun replied:
Speaking of `NamedTemporaryFile()`, handling the case where creating the wrapper fails has to be updated for `delete_on_close` as well. For example:

```python
    except:
        if name is not None and not (
          _os.name == 'nt' and delete and delete_on_close):
            _os.unlink(name)
        raise
```

ezio-melotti commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,53 @@ The module defines the following user-callable items:
+     the directory entry is not unlinked).
+
+   * There is more granularity in the deletion behaviour of the file (see
+     ``delete_on_close`` below)
```
```suggestion
   * The file is guaranteed to have a visible name in the file system
     (on Unix, the directory entry is not unlinked).
   * There is more granularity in the deletion behaviour of the file
     (see *delete_on_close* below)
```
Fixed *delete_on_close* markup, rewrapped, removed newline.

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -979,6 +979,88 @@ def test_del_on_close(self):
 
+    def test_not_del_on_close_if_delete_on_close_false(self):
+        # A NamedTemporaryFile is NOT deleted when closed if
+        # delete_on_close= False, but is deleted on content manager exit
```
```suggestion
        # delete_on_close=False, but is deleted on context manager exit
```

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -979,6 +979,88 @@ def test_del_on_close(self):
+                    # Testing that file is not deleted on close
+                    self.assertTrue(os.path.exists(f.name),
+                            "NamedTemporaryFile %s is incorrectly deleted\
+                             on closure when delete_on_close= False" % f_name)
```
```suggestion
                    self.assertTrue(os.path.exists(f.name),
                            f"NamedTemporaryFile {f.name!r} is incorrectly deleted "
                            f"on closure when delete_on_close=False")
```
I switched to f-strings to fix the output message, `f.name` (instead of `f_name`) for consistency, and added `!r`.

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -979,6 +979,88 @@ def test_del_on_close(self):
+                # Testing that file is deleted on content manager exit
+                self.assertFalse(os.path.exists(f.name),
+                                 "NamedTemporaryFile %s exists\
+                                  after content manager exit" % f.name)
```
```suggestion
                self.assertFalse(os.path.exists(f.name),
                                 f"NamedTemporaryFile {f.name!r} exists "
                                 f"after context manager exit")
```

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -979,6 +979,88 @@ def test_del_on_close(self):
+                             on closure when delete_on_close= False" % f_name)
+
+            with self.subTest():
+                # Testing that file is deleted on content manager exit
```
```suggestion
                # Testing that file is deleted on context manager exit
```

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -979,6 +979,88 @@ def test_del_on_close(self):
+
+    def test_ok_to_delete_manually(self):
+        # A NamedTemporaryFile can be deleted by a user before content manager
+        # comes to it. This will not generate an error
```
```suggestion
        # A NamedTemporaryFile can be deleted by a user before
        # context manager comes to it. This will not generate an error
```

Ev2geny committed:
Deletion of file upon finalization is moved from _TemporaryFileWrapper to _TemporaryFileCloser
Test test_del_by_finalizer_if_no_with is changed from simulating finalizer by calling __del__() to making
sure finalizer really runs

eryksun commented on Lib/tempfile.py:
```diff
@@ -446,16 +446,19 @@ def close(self, unlink=_os.unlink):
+        # Need to ensure the file is deleted on __del__ if delete = True and
+        # file still exists
+        if self.delete and _os.path.exists(self.name):
+            _os.unlink(self.name)
```
Implementing this here is a mistake. Deleting the file in `_TemporaryFileCloser` should only be due to `delete_on_close`. It should not try to delete the file if it was already deleted. We don't know whether an unrelated file was created with the same name. In particular, `_TemporaryFileCloser.__del__` should not try to delete the file again if it's handled by `close()`, i.e. if `delete_on_close` is true, and it should not try to delete the file again if `__exit__()` was called. But how does `_TemporaryFileCloser` know whether `_TemporaryFileWrapper.__exit__()` was called? The file being closed is not enough information, since `close()` may have been called manually. It is not the job of `_TemporaryFileCloser` to clean up the file if `delete` is true and `delete_on_close` is false. That's the job of `_TemporaryFileWrapper`.

Ev2geny replied:
Thanks for spending time by looking at this PR.
I acknowledge I moved the code to a different place, than you suggested, but I saw, that `__del__()` is already captured in the `_TemporaryFileCloser` (see [here ](https://github.com/python/cpython/blob/80bc7d7c0a3f777da244ec79429baf382523b8da/Lib/tempfile.py#L448))

I also noticed, that `__del__()` in  `_TemporaryFileCloser` is called after the `__del__()` on `_TemporaryFileWrapper` (which was creating a problem with the tests, when run on Linux, which are gone now).

Any way, in your suggestion, something will have to be done with existing `__del__` code in the `_TemporaryFileCloser` (see [here ](https://github.com/python/cpython/blob/80bc7d7c0a3f777da244ec79429baf382523b8da/Lib/tempfile.py#L448)) to avoid the problem in POSIX

eryksun replied:
`_TemporaryFileCloser.__del__()` is only there to ensure that the file gets closed. This also guarantees to delete the file if both `delete` and `delete_on_close` are true. In this case, `unlink()` is guaranteed to only get called once because the work is gated by `close_called`. This behavior must not change. 

Similarly, if `delete` is true and `delete_on_close` is false, if the file is deleted by `__exit__()`, then `_TemporaryFileWrapper.__del__()` must not try to delete the file again. That's why I added the `_cleanup_called` class attribute in my example and removed redundant work by adding a common `_cleanup()` method.

The combined behavior of `_TemporaryFileCloser` and `_TemporaryFileWrapper` must never call `unlink()` more than once.

Ev2geny replied:
@eryksun , thanks!

I understand, that the combined behavior of `_TemporaryFileCloser` and `_TemporaryFileWrapper` must never call `unlink()` more than once. 

What I still don't get is the following: in the current Python release we already have a code, which handles finalization event in the `_TemporaryFileCloser`.  If we now add another `__del__()` into the  `_TemporaryFileWrapper`, then during finalization first the `_TemporaryFileWrapper.__del__()` will run and later the `_TemporaryFileCloser.__del__()` and we have to make sure they don't conflict with each other plus that it is clear for anyone how it all works together with each other

Isn't it is better from design and readability perspective to place a code, which handles the finalization event into one place and not spread it across 2 classes?

eryksun replied:
There is no conflict, and the behavior should be simple to understand. The closer only deletes the file if both `delete` and `delete_on_close` are true. The wrapper only deletes the file if `delete` is true and `delete_on_close` is false. Both cases need to be gated by variables such as `close_called` and `_cleanup_called` in order to ensure that `unlink()` is only called once. In particular, when `delete` and `delete_on_close` are true, the finalizer of the closer shouldn't try to delete the file if `close()` was already called, and if `delete` is true and `delete_on_close` is false, the finalizer of the wrapper shouldn't try to delete the file if `__exit__()` was already called.

The closer on its own cannot properly handle the case of `delete` is true and `delete_on_close` is false. Just by the fact that `delete_on_close` is false, you know that it's not the job of the closer to handle this case. In particular, the closer doesn't know, and shouldn't know, whether `__exit__()` was called in the wrapper. So it doesn't know whether `unlink()` was already called. A new file could have been created with the same name after `__exit__()` was called, in which case calling `unlink()` again would be seriously wrong.

Ev2geny replied:
OK, understood, thanks.
Another, a bit different, question: where the situation will be handled, in case `delete = True` and `delete_on_close = False` and no context manager was used and file was not even explicitly closed (not a good programming).

Like this:

```py
        def my_func(dir)->str:
            f = tempfile.NamedTemporaryFile(dir=dir, delete=True, delete_on_close=False)
            tmp_name = f.name
            f.write(b'blat')
            return tmp_name
```
Which part of the code will be responsible to delete  the file?

eryksun replied:
The file would be deleted by the wrapper's `__del__()` method, which in the example I wrote would call the `_cleanup()` method. In CPython, the delete will occur immediately when `my_func()` returns, so I can't imagine what reason anyone would have to return `tmp_name`, the name of a deleted file.

Ev2geny replied:
@eryksun , sorry if this is a stupid question, but can you explain me please why we need **class** attributes `name = None` and `_cleanup_called = False`, why can these not be an instance attributes, initialised in `__init__` ?

In all the documentation I read they say class attributes are used if data needs to be shared by all instances. Surely the flag `_cleanup_called` we want to be specific for every instance of the temporary file, as we can have many of them (or am I badly missing something)? 

I see the similar approach also being used in the `_TemporaryFileCloser`

Ev2geny replied:
I have now implemented this as per your suggestions . There are 2 things however

- I still have some questions about usage of class attributes (see above)

- I had to add one if statement [here](https://github.com/Ev2geny/cpython/blob/5de68c0556058c762572e8afe4c2981b38987a1b/Lib/tempfile.py#L447):

```py
                    if self.delete and self.delete_on_close:
                        if _os.path.exists(self.name):
                            unlink(self.name)
```

Otherwise the tests were not passing for test_tempfile
- There are some errors now introduced in other parts of Python. So I think will need your advice what could be wrong

<details>

  <summary>Click me</summary>

== CPython 3.12.0a0 (heads/fix-issue-14243:0e0b6ba04d, Sep 23 2022, 11:54:11) [GCC 11.2.0]

== Linux-5.15.0-48-generic-x86_64-with-glibc2.35 little-endian

== cwd: /home/evg/evpy/cpython/build/test_python_6764æ

== CPU count: 3

== encodings: locale=UTF-8, FS=utf-8

0:00:00 load avg: 0.30 Run tests sequentially

0:00:00 load avg: 0.30 [1/1] test_tempfile

test_nonempty_list (test.test_tempfile.TestCandidateTempdirList.test_nonempty_list) ... ok

test_wanted_dirs (test.test_tempfile.TestCandidateTempdirList.test_wanted_dirs) ... ok

test_exports (test.test_tempfile.TestExports.test_exports) ... ok

test_retval (test.test_tempfile.TestGetCandidateNames.test_retval) ... ok

test_same_thing (test.test_tempfile.TestGetCandidateNames.test_same_thing) ... ok

test_no_files_left_behind (test.test_tempfile.TestGetDefaultTempdir.test_no_files_left_behind) ... ok

test_case_sensitive (test.test_tempfile.TestGetTempDir.test_case_sensitive) ... ok

test_directory_exists (test.test_tempfile.TestGetTempDir.test_directory_exists) ... ok

test_directory_writable (test.test_tempfile.TestGetTempDir.test_directory_writable) ... ok

test_same_thing (test.test_tempfile.TestGetTempDir.test_same_thing) ... ok

test_sane_template (test.test_tempfile.TestGetTempPrefix.test_sane_template) ... ok

test_usable_template (test.test_tempfile.TestGetTempPrefix.test_usable_template) ... ok

test_infer_return_type_multiples (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_multiples) ... ok

test_infer_return_type_multiples_and_none (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_multiples_and_none) ... ok

test_infer_return_type_pathlib (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_pathlib) ... ok

test_infer_return_type_pathlike (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_pathlike) ... ok

test_infer_return_type_singles (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_singles) ... ok

test_basic (test.test_tempfile.TestMkdtemp.test_basic) ... ok

test_basic_many (test.test_tempfile.TestMkdtemp.test_basic_many) ... ok

test_basic_with_bytes_names (test.test_tempfile.TestMkdtemp.test_basic_with_bytes_names) ... ok

test_choose_directory (test.test_tempfile.TestMkdtemp.test_choose_directory) ... ok

test_collision_with_existing_directory (test.test_tempfile.TestMkdtemp.test_collision_with_existing_directory) ... ok

test_collision_with_existing_file (test.test_tempfile.TestMkdtemp.test_collision_with_existing_file) ... ok

test_for_tempdir_is_bytes_issue40701_api_warts (test.test_tempfile.TestMkdtemp.test_for_tempdir_is_bytes_issue40701_api_warts) ... ok

test_mode (test.test_tempfile.TestMkdtemp.test_mode) ... ok

test_non_directory (test.test_tempfile.TestMkdtemp.test_non_directory) ... ok

test_nonexisting_directory (test.test_tempfile.TestMkdtemp.test_nonexisting_directory) ... ok

test_read_only_directory (test.test_tempfile.TestMkdtemp.test_read_only_directory) ... ok

test_basic (test.test_tempfile.TestMkstemp.test_basic) ... ok

test_basic_with_bytes_names (test.test_tempfile.TestMkstemp.test_basic_with_bytes_names) ... ok

test_choose_directory (test.test_tempfile.TestMkstemp.test_choose_directory) ... ok

test_for_tempdir_is_bytes_issue40701_api_warts (test.test_tempfile.TestMkstemp.test_for_tempdir_is_bytes_issue40701_api_warts) ... ok

test_basic (test.test_tempfile.TestMkstempInner.test_basic) ... ok

test_basic_many (test.test_tempfile.TestMkstempInner.test_basic_many) ... ok

test_basic_with_bytes_names (test.test_tempfile.TestMkstempInner.test_basic_with_bytes_names) ... ok

test_choose_directory (test.test_tempfile.TestMkstempInner.test_choose_directory) ... ok

test_collision_with_existing_directory (test.test_tempfile.TestMkstempInner.test_collision_with_existing_directory) ... ok

test_collision_with_existing_file (test.test_tempfile.TestMkstempInner.test_collision_with_existing_file) ... ok

test_file_mode (test.test_tempfile.TestMkstempInner.test_file_mode) ... ok

test_noinherit (test.test_tempfile.TestMkstempInner.test_noinherit) ... ok

test_non_directory (test.test_tempfile.TestMkstempInner.test_non_directory) ... ok

test_nonexisting_directory (test.test_tempfile.TestMkstempInner.test_nonexisting_directory) ... ok

test_read_only_directory (test.test_tempfile.TestMkstempInner.test_read_only_directory) ... ok

test_textmode (test.test_tempfile.TestMkstempInner.test_textmode) ... skipped 'text mode not available'

test_basic (test.test_tempfile.TestMktemp.test_basic) ... ok

test_many (test.test_tempfile.TestMktemp.test_many) ... ok

test_bad_encoding (test.test_tempfile.TestNamedTemporaryFile.test_bad_encoding) ... ok

test_bad_mode (test.test_tempfile.TestNamedTemporaryFile.test_bad_mode) ... ok

test_basic (test.test_tempfile.TestNamedTemporaryFile.test_basic) ... ok

test_context_man_not_del_if_delete_false (test.test_tempfile.TestNamedTemporaryFile.test_context_man_not_del_if_delete_false) ... ok

test_context_man_not_del_on_close_if_delete_on_close_false (test.test_tempfile.TestNamedTemporaryFile.test_context_man_not_del_on_close_if_delete_on_close_false) ... ok

test_context_man_ok_to_delete_manually (test.test_tempfile.TestNamedTemporaryFile.test_context_man_ok_to_delete_manually) ... ok

test_context_manager (test.test_tempfile.TestNamedTemporaryFile.test_context_manager) ... ok

test_correct_finalizer_work_if_already_deleted (test.test_tempfile.TestNamedTemporaryFile.test_correct_finalizer_work_if_already_deleted) ... ok

test_creates_named (test.test_tempfile.TestNamedTemporaryFile.test_creates_named) ... ok

test_del_by_finalizer (test.test_tempfile.TestNamedTemporaryFile.test_del_by_finalizer) ... ok

test_del_on_close (test.test_tempfile.TestNamedTemporaryFile.test_del_on_close) ... ok

test_dis_del_on_close (test.test_tempfile.TestNamedTemporaryFile.test_dis_del_on_close) ... ok

test_iter (test.test_tempfile.TestNamedTemporaryFile.test_iter) ... ok

test_method_lookup (test.test_tempfile.TestNamedTemporaryFile.test_method_lookup) ... ok

test_multiple_close (test.test_tempfile.TestNamedTemporaryFile.test_multiple_close) ... ok

test_unexpected_error (test.test_tempfile.TestNamedTemporaryFile.test_unexpected_error) ... ok

test_get_eight_char_str (test.test_tempfile.TestRandomNameSequence.test_get_eight_char_str) ... ok

test_many (test.test_tempfile.TestRandomNameSequence.test_many) ... ok

test_process_awareness (test.test_tempfile.TestRandomNameSequence.test_process_awareness) ... ok

test_basic (test.test_tempfile.TestSpooledTemporaryFile.test_basic) ... ok

test_bound_methods (test.test_tempfile.TestSpooledTemporaryFile.test_bound_methods) ... ok

test_class_getitem (test.test_tempfile.TestSpooledTemporaryFile.test_class_getitem) ... ok

test_context_manager_after_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_context_manager_after_rollover) ... ok

test_context_manager_before_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_context_manager_before_rollover) ... ok

test_context_manager_during_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_context_manager_during_rollover) ... ok

test_del_on_close (test.test_tempfile.TestSpooledTemporaryFile.test_del_on_close) ... ok

test_del_rolled_file (test.test_tempfile.TestSpooledTemporaryFile.test_del_rolled_file) ... ok

test_del_unrolled_file (test.test_tempfile.TestSpooledTemporaryFile.test_del_unrolled_file) ... ok

test_fileno (test.test_tempfile.TestSpooledTemporaryFile.test_fileno) ... ok

test_iobase_interface (test.test_tempfile.TestSpooledTemporaryFile.test_iobase_interface) ... ok

test_is_iobase (test.test_tempfile.TestSpooledTemporaryFile.test_is_iobase) ... ok

test_multiple_close_after_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_multiple_close_after_rollover) ... ok

test_multiple_close_before_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_multiple_close_before_rollover) ... ok

test_properties (test.test_tempfile.TestSpooledTemporaryFile.test_properties) ... ok

test_rewrite_small (test.test_tempfile.TestSpooledTemporaryFile.test_rewrite_small) ... ok

test_sparse (test.test_tempfile.TestSpooledTemporaryFile.test_sparse) ... ok

test_text_mode (test.test_tempfile.TestSpooledTemporaryFile.test_text_mode) ... ok

test_text_newline_and_encoding (test.test_tempfile.TestSpooledTemporaryFile.test_text_newline_and_encoding) ... ok

test_truncate_with_size_parameter (test.test_tempfile.TestSpooledTemporaryFile.test_truncate_with_size_parameter) ... ok

test_write_sequential (test.test_tempfile.TestSpooledTemporaryFile.test_write_sequential) ... ok

test_writelines (test.test_tempfile.TestSpooledTemporaryFile.test_writelines) ... ok

test_writelines_sequential (test.test_tempfile.TestSpooledTemporaryFile.test_writelines_sequential) ... ok

test_cleanup_with_symlink_to_a_directory (test.test_tempfile.TestTemporaryDirectory.test_cleanup_with_symlink_to_a_directory) ... ok

test_context_manager (test.test_tempfile.TestTemporaryDirectory.test_context_manager) ... ok

test_del_on_collection (test.test_tempfile.TestTemporaryDirectory.test_del_on_collection) ... ok

test_del_on_collection_ignore_errors (test.test_tempfile.TestTemporaryDirectory.test_del_on_collection_ignore_errors)

Test that ignoring errors works when TemporaryDirectory is gced. ... ok

test_del_on_shutdown (test.test_tempfile.TestTemporaryDirectory.test_del_on_shutdown) ... ok

test_del_on_shutdown_ignore_errors (test.test_tempfile.TestTemporaryDirectory.test_del_on_shutdown_ignore_errors)

Test ignoring errors works when a tempdir is gc'ed on shutdown. ... ok

test_exit_on_shutdown (test.test_tempfile.TestTemporaryDirectory.test_exit_on_shutdown) ... ok

test_explicit_cleanup (test.test_tempfile.TestTemporaryDirectory.test_explicit_cleanup) ... ok

test_explict_cleanup_ignore_errors (test.test_tempfile.TestTemporaryDirectory.test_explict_cleanup_ignore_errors)

Test that cleanup doesn't return an error when ignoring them. ... ok

test_flags (test.test_tempfile.TestTemporaryDirectory.test_flags) ... skipped 'requires os.lchflags'

test_mkdtemp_failure (test.test_tempfile.TestTemporaryDirectory.test_mkdtemp_failure) ... ok

test_modes (test.test_tempfile.TestTemporaryDirectory.test_modes) ... ok

test_multiple_close (test.test_tempfile.TestTemporaryDirectory.test_multiple_close) ... ok

test_warnings_on_cleanup (test.test_tempfile.TestTemporaryDirectory.test_warnings_on_cleanup) ... ok

test_bad_encoding (test.test_tempfile.TestTemporaryFile.test_bad_encoding) ... ok

test_bad_mode (test.test_tempfile.TestTemporaryFile.test_bad_mode) ... ok

test_basic (test.test_tempfile.TestTemporaryFile.test_basic) ... ok

test_has_no_name (test.test_tempfile.TestTemporaryFile.test_has_no_name) ... ok

test_mode_and_encoding (test.test_tempfile.TestTemporaryFile.test_mode_and_encoding) ... ok

test_multiple_close (test.test_tempfile.TestTemporaryFile.test_multiple_close) ... ok

test_unexpected_error (test.test_tempfile.TestTemporaryFile.test_unexpected_error) ... ok



----------------------------------------------------------------------

Ran 109 tests in 1.041s



OK (skipped=2)

test_tempfile failed (env changed)



== Tests result: SUCCESS ==



1 test altered the execution environment:

    test_tempfile



Total duration: 1.2 sec

Tests result: SUCCESS

</details>

Ev2geny committed:
Some additions to Lib/test/test_tempfile.py

Ev2geny committed:
clearing of the files in the situation when delete and not delete_on_close is moved to _TemporaryFileWrapper
based on the advice from @eryksun

There are errors in the test (at least on Linux)

Ev2geny committed:
if statement is added to avoid errors, when testing. Not sure 100% is this is a solution or workaround

eryksun commented on Lib/tempfile.py:
```diff
@@ -421,10 +421,11 @@ class _TemporaryFileCloser:
     close_called = False
 
-    def __init__(self, file, name, delete=True):
+    def __init__(self, file, name, delete=True, delete_on_close=True):
```
Nothing needs to change in this class. Instead, change the way it gets instantiated in `_TemporaryFileWrapper.__init__()` as follows: `_TemporaryFileCloser(file, name, delete and delete_on_close)`. The meaning of `delete` in the closer has always been "delete on close", and that won't change.

Ev2geny replied:
This is now implemented in[ this commit](https://github.com/python/cpython/pull/97015/commits/6944ab23e513861508fe643558efce47f178875e)

eryksun commented on Lib/tempfile.py:
```diff
@@ -495,13 +501,31 @@ def __enter__(self):
         self.file.__enter__()
         return self
 
+    def _cleanup(self):
```
I forgot to cache the `unlink()` function locally for the case when the `os` module is cleared at shutdown:

```suggestion
    def _cleanup(self, unlink=_os.unlink):
```

eryksun commented on Lib/tempfile.py:
```diff
@@ -495,13 +501,31 @@ def __enter__(self):
+            # If the file is to be deleted, but not on close, delete it now.
+            if self.delete and not self.delete_on_close:
+                try:
+                    _os.unlink(self.name)
```
```suggestion
                    unlink(self.name)
```

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+            tmp_name = my_func(dir)
+            self.assertFalse(os.path.exists(tmp_name),
+                        "NamedTemporaryFile %s exists after finalizer "\
+                            % tmp_name)
```
```suggestion
            self.assertFalse(os.path.exists(tmp_name),
                        f"NamedTemporaryFile {tmp_name!r} exists after finalizer ")
```

ezio-melotti commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+                f.write(b'blat')
+                f_name = f.name
+            self.assertTrue(os.path.exists(f.name),
+                        "NamedTemporaryFile %s exists after close" % f.name)
```
```suggestion
            self.assertTrue(os.path.exists(f.name),
                        f"NamedTemporaryFile {f.name!r} exists after close")
```

eryksun commented on Lib/tempfile.py:
```diff
@@ -518,18 +542,32 @@ def __iter__(self):
+            # been deleted already.
+            except FileNotFoundError:
+                pass
+
```
The finalizer should simply call `self._cleanup()`. 

```suggestion
        self._cleanup()
```
Note that `_cleanup()` is guarded by `_cleanup_called`, so it only does the work the first time it's called.  Also, note that `_cleanup()` first closes the file before trying to delete it. This is required because `os.open()` without `O_TEMPORARY` does not share `DELETE` access.

Ev2geny committed:
Apply suggestions from code review

Implemented review from @eryksun and @merwok

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>
Co-authored-by: Eryk Sun <eryksun@gmail.com>
Co-authored-by: Éric <merwok@netwok.org>

Ev2geny committed:
__del__ was defined twice in _TemporaryFileWrapper. Fixed now

Ev2geny committed:
_TemporaryFileCloser is returned to original state and the logic, which takes into account delete_on_close moved to _TemporaryFileWrapper (suggestion of @ekisun)

Ev2geny committed:
All delete operations are moved to the class _TemporaryFileCloser
	modified:   Lib/tempfile.py

Ev2geny committed:
Minor update to comments

Ev2geny committed:
Whitespec fixes by patchcheck.py
	modified:   Lib/tempfile.py

eryksun commented on Lib/tempfile.py:
```diff
@@ -418,44 +418,40 @@ class _TemporaryFileCloser:
 
-
+    def __del__(self):
+        self.cleanup()
```
```suggestion
        self.cleanup()


```

eryksun commented on Lib/tempfile.py:
```diff
@@ -418,44 +418,40 @@ class _TemporaryFileCloser:
-                finally:
-                    if self.delete:
+            finally:
+                if self.delete and not (self.delete_on_close and windows):
```
```suggestion
                # Windows provides delete-on-close as a primitive, in which
                # case the file was deleted by self.file.close().
                if self.delete and not (self.delete_on_close and windows):
```

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Modified exception handling in NamedTemporaryFile

	modified:   Lib/tempfile.py

Ev2geny committed:
Wrapping at 80 characters is fixed

	modified:   Lib/test/test_tempfile.py

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,52 @@ The module defines the following user-callable items:
+   following differences:
+
+   * The file is guaranteed to have a visible name in the file system
+     (on Unix, the directory entry is not unlinked).
```
I know this line was here before, but it doesn't make any sense to me. Should we mention this somewhere else?

Ev2geny replied:
@eryksun , I think **zooba** talks about the line "(on Unix, the directory entry is not unlinked)." What do you think, does it make scene to move it somewhere else?

eryksun replied:
On POSIX, a file created by `TemoraryFile()` does not have a visible name in the filesystem. The implementation tries to use the `O_TMPFILE` flag if it's supported (Linux 3.11+), which creates an anonymous file. If `O_TMPFILE` isn't supported, the file is created with a unique name, and then it's immediately unlinked while the file descriptor is still open.

On Windows, `TemporaryFile` is aliased to `NamedTemporaryFile`. There isn't a compelling reason for this decision. It's fine to delete a file that's opened with the `O_TEMPORARY` flag. On Windows 10 and 11, if it's on an NTFS filesystem, the file gets unlinked immediately. Prior to Windows 10, or on filesystems other than NTFS, the file is deleted but not unlinked. In the latter state, a file has a visible entry in the directory, but it can't be opened again for any access, and it gets unlinked as soon as the last handle to it is closed.

Ev2geny replied:
@eryksun , thanks for explanation. So, would you agree, that the statement _on Unix, the directory entry is not unlinked_ does not really make too much scene the way it is written now? Would it be better to either remove it completely or to put a broader explanation, similar to what you provided?

eryksun replied:
It's in the context of clarifying that `NamedTemporaryFile()` acts "exactly" like `TemporaryFile()`. In that context it makes sense. But I'd prefer that the first sentence and the two bullet points were replaced with something like the following:

```text
   This function returns a file that is guaranteed to have a visible name in
   the file system. To manage the named file, it extends the parameters of
   :func:`TemporaryFile` with *delete* and *delete_on_close* parameters that
   determine whether and how the named file should be automatically deleted.
```

Ev2geny replied:
I have now placed the statement about unlinking on Unix to a different place and slightly reworked it to make a reference to TemporaryFile(), without which I agree it was a bit confusing.

I have also implemented rewording from @eryksun

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,52 @@ The module defines the following user-callable items:
+   only, if no context manager was used, then the file is deleted only when the
+   :term:`file-like object` is finalized and hence deletion is not always
+   guaranteed in this case (see :meth:`object.__del__`). If *delete* is false,
+   the value of *delete_on_close* is ignored.
```
```suggestion
   only, or else when the :term:`file-like object` is finalized. Deletion is not
   always guaranteed in this case (see :meth:`object.__del__`). If *delete* is
   false, the value of *delete_on_close* is ignored.
```

Ev2geny replied:
I have now implemented this locally on my PC.

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,52 @@ The module defines the following user-callable items:
+   Windows, if *delete* is true and *delete_on_close* is false, additional
+   opens that do not share delete access (e.g. via builtin :func:`open`) must
+   be closed before exiting the context manager, else the :func:`os.unlink`
+   call on context manager exit will fail with a :exc:`PermissionError`.
```
Can we reframe this as:

* the file can be opened again using its name
* on Windows, subsequent opens should share delete access (e.g.)
* however, when `delete` or `delete_on_close` are false, sharing delete access is not necessary, but you should ensure the file is closed before leaving the context manager

This sets it up more positively for the user - what _should_ they do, as opposed to what could go wrong.

Ev2geny replied:
Please check how I have written this in the new version. Not sure everybody would agree, but I have now written this information the way I would want it to be explained for me without the risk of twisting my brains.

**Additional question:** does it make sense to  write unit tests to cover all these combinations of opening the file the second time, while it is already open.  I have tested this for myself (on Windows) to make sure it is all correct, but I am not sure that it shall go to unit tests.

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,52 @@ The module defines the following user-callable items:
+   parameter to be false) or, in case the temporary file is created in a
+   :keyword:`with` statement, set the *delete_on_close* parameter to be false.
+   The latter approach is recommended as it provides assistance in automatic
+   cleaning of the temporary file upon the context manager exit.
```
This paragraph seems redundant with the previous one?

Ev2geny replied:
The previous one talks about opening the file, while it is still open, This section talks about opening the file after it has already been closed, which is a bit different case. I agree one can work it out from the information at the beginning of the section, but I thought that because it represents arguably the most frequent use, then may be it would not harm to have this paragraph.

Any way, I have now moved it to a slightly different place and have re-written to make sure a reader understands, that this is not a new information, but the practical conclusion from the previous paragraph.

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -346,6 +381,22 @@ Here are some examples of typical usage of the :mod:`tempfile` module::
+    # the file is closed, but not removed
+    # open the file again by using its name
+    ...     f = open(fp_name)
+    ...     f.seek(0)
```
We shouldn't need to seek, right?

Ev2geny replied:
Correct, I removed it now

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -346,6 +381,22 @@ Here are some examples of typical usage of the :mod:`tempfile` module::
+    ...     fp.close()
+    # the file is closed, but not removed
+    # open the file again by using its name
+    ...     f = open(fp_name)
```
`fp.name` is still valid here, right? Why capture it?

zooba replied:
Also, maybe use this one in its own nested `with` block?

Ev2geny replied:
I captured the `fp.name` to make more explicit, but I agree, it is redundant, so I removed it now.
Also nested `with` block is added.

zooba commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+    def test_del_by_finalizer(self):
+        # A NamedTemporaryFile is deleted by fanalizer in case delete = True
+        # delete_on_close = False and no context manager is used
+        def my_func(dir)->str:
```
Spaces around the arrow if you're going to do it, or just leave it out. The function isn't going to travel that far.

Ev2geny replied:
Type notation is removed

zooba reviewed (commented):
Code looks good to me. Would like to make the docs easier to follow while we're here though.

eryksun commented on Lib/tempfile.py:
```diff
@@ -418,42 +418,42 @@ class _TemporaryFileCloser:
+            finally:
+                # Windows provides delete-on-close as a primitive, in which
+                # case the file was deleted by self.file.close().
+                if self.delete and not (self.delete_on_close and windows):
```
Let's short-circuit to false on POSIX:

```suggestion
                if self.delete and not (windows and self.delete_on_close):
```

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Documentation update, based on the feedback from @zooba
	modified:   Doc/library/tempfile.rst

Ev2geny committed:
Minor PEP8 update based on the feedback from @zooba
	modified:   Lib/test/test_tempfile.py

eryksun commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+
+    def test_del_by_finalizer(self):
+        # A NamedTemporaryFile is deleted by fanalizer in case delete = True
+        # delete_on_close = False and no context manager is used
```
```suggestion
        # A NamedTemporaryFile is deleted when finalized in the case of
        # delete=True, delete_on_close=False, and no with-statement is used.
```

eryksun commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+            f.close()
+            os.unlink(tmp_name)
+            return tmp_name
+        # Making sure that Garbage Collector has finalized the file object
```
```suggestion
        # Make sure that the garbage collector has finalized the file object.
```

eryksun commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+    def test_correct_finalizer_work_if_already_deleted(self):
+        # There should be No errors in case delete = True
+        # delete_on_close = False and no context manager is used, but file is
+        # deleted manually
```
```suggestion
        # There should be no error in the case of delete=True,
        # delete_on_close=False, no with-statement is used, and the file is
        # deleted manually.
```

eryksun commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+            # Testing extreme case, where the file is not explicitly closed
+            # f.close()
+            return tmp_name
+        # Making sure that Garbage Collector has finalized the file object
```
```suggestion
        # Make sure that the garbage collector has finalized the file object.
```

eryksun commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+        f_name = ""
+        try:
+            # setting delete_on_close = True to test, that this does not have
+            # an effect, if delete = False
```
```suggestion
            # Test that delete_on_close=True has no effect if delete=False.
```

eryksun commented on Lib/test/test_tempfile.py:
```diff
@@ -1013,6 +1014,103 @@ def use_closed():
+
+    def test_context_man_ok_to_delete_manually(self):
+        # A NamedTemporaryFile can be deleted by a user before
+        # context manager comes to it. This will not generate an error
```
```suggestion
        # In the case of delete=True, a NamedTemporaryFile can be manually
        # deleted in a with-statement context without causing an error.
```

Ev2geny committed:
Improving comments in the init test file, based on the feedback from @eryksun

Files changed:
      Lib/test/test_tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Updating of documentation

	modified:   library/tempfile.rst

Ev2geny committed:
Merge branch 'fix-issue-14243' of https://github.com/Ev2geny/cpython into fix-issue-14243

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -80,41 +80,50 @@ The module defines the following user-callable items:
+   false, the value of *delete_on_close* is ignored.
+
+   Therefore to use the name of the temporary file to open the closed file
+   second time, either make sure not to delete the file upon closure (set the
```
```suggestion
   Therefore to use the name of the temporary file to reopen the file after
   closing it, either make sure not to delete the file upon closure (set the
```

zooba commented on Doc/library/tempfile.rst:
```diff
@@ -80,41 +80,50 @@ The module defines the following user-callable items:
+
+   * On POSIX the file can always be opened again.
+   * On Windows to open the file again while it is open make sure that at least
+     one of the following conditions are fulfilled:
```
```suggestion
   * On Windows, make sure that at least one of the following conditions are
     fulfilled:
```

zooba committed:
Apply suggestions from code review
